extensions/coap/protocols/CoapC2Protocol.cpp

```diff
@@ -0,0 +1,353 @@
+      }
+      break;
+    case HEARTBEAT:
+      payload_type = 1;
+      stream.write(&payload_type, 1);
+      if (writeHeartbeat(&stream, payload) != 0) {
+        return C2Payload(payload.getOperation(), state::UpdateState::READ_ERROR, true);
+      }
+      break;
+  };
+
+  size_t bsize = stream.getSize();
+  auto buffer = (unsigned char*) stream.getBuffer();
+
+  controllers::CoapConnectorService::CoAPMessage message(coap_service_->sendPayload(COAP_REQUEST_POST, endpoint, buffer, bsize));
+
+  if (message.isRegistrationRequest()) {
+    require_registration_ = true;
+  } else if (message.getSize() > 0) {
+    io::DataStream byteStream(message.getData(), message.getSize());
+    io::BaseStream responseStream(&byteStream);
+    responseStream.read(version);
+    responseStream.read(size);
+    C2Payload new_payload(payload.getOperation(), state::UpdateState::NESTED, true);
+    for (int i = 0; i < size; i++) {
+
+      uint8_t operationType;
+      uint16_t argsize = 0;
+      std::string operand, id;
+      responseStream.read(operationType);
+      responseStream.readUTF(id, false);
+      responseStream.readUTF(operand, false);
+
+      auto newOp = getOperation(operationType);
+      C2Payload nested_payload(newOp, state::UpdateState::READ_COMPLETE, true);
+      nested_payload.setIdentifier(id);
+      C2ContentResponse new_command(newOp);
+      new_command.delay = 0;
+      new_command.required = true;
+      new_command.ttl = -1;
+      new_command.name = operand;
+      new_command.ident = id;
+      responseStream.read(argsize);
+      for (int j = 0; j < argsize; j++) {
+        std::string key, value;
+        responseStream.readUTF(key);
+        responseStream.readUTF(value);
+        new_command.operation_arguments[key] = value;
+      }
+
+      nested_payload.addContent(std::move(new_command));
+      new_payload.addPayload(std::move(nested_payload));
+    }
+    return new_payload;
+  }
+
+  return C2Payload(payload.getOperation(), state::UpdateState::READ_ERROR, true);
+}
+
+const C2Payload * const CoapProtocol::getPayload(const std::string &name, const C2Payload * const payload) {
+  if (payload->getLabel() == name) {
+    return payload;
+  }
+  for (const auto &nested_payload : payload->getNestedPayloads()) {
+    if (nested_payload.getLabel() == name) {
+      return &nested_payload;
+    }
+  }
+  return nullptr;
+}
+
+int8_t CoapProtocol::getLong(const C2ContentResponse * const cmp, const std::string &name, uint64_t *value) {
+  auto exists = cmp->operation_arguments.find(name);
+  if (exists != cmp->operation_arguments.end()) {
+    const auto &node = exists->second.getValue();
+    if (auto sub_type = std::dynamic_pointer_cast<state::response::Int64Value>(node)) {
+      *value = sub_type->getValue();
+      return 0;
+    }
+  }
+  return -1;
+}
+
+int8_t CoapProtocol::getString(const C2ContentResponse * const cmp, const std::string &name, std::string *value) {
```

Comment:
As it's C++, why value is a pointer (ref would be fine)?
Moreover I don't see why to have 0 or -1 return codes, empty string would be fine as well to indicate the lack of existence.